### PR TITLE
Add polyfill for console

### PR DESCRIPTION
**Pull request by whitneyit**
This PR adds a poly fill for `console` and its various methods. RE: #440 

Whilst putting the polyfill together I was unable to come up with a definite list of methods that each browsers support and in which version they support which method :cry:

Also, I'm not entirely sure how we detect each missing method. Would splitting each method into its own folder similar to [Array](https://github.com/Financial-Times/polyfill-service/tree/master/polyfills/Array) work better?

### CODE REVIEW

File: polyfills/console/tests.js [hunk 1]
```diff
@@ -0,0 +1,181 @@
+describe('console', function () {
+
+	it('should expose itself on the global Object', function () {
+		expect(console).to.be.ok;
+	});
+
+	describe('[methods]', function () {
```
Comment (triblondon) [inline polyfills/console/tests.js]: props

File: polyfills/console/polyfill.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+		// Check if we can safely use `fn.apply`.
+		if (hasApply) {
+			return function () {
+				fn.apply(console, arguments);
+			};
+		}
+
+		// If we are here it means that `console.log` does not have an
+		// `fn.apply` method so we log out the first 3 arguments.
```
Comment (triblondon) [inline polyfills/console/polyfill.js]: Not necessarily the log method.

File: polyfills/console/polyfill.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		// Reading `fn.apply` throws in IE11  in IE8 mode. This is because
+		// `typeof console.log === "object"` in IE8
+		try {
+			hasApply = !!fn.apply;
+		} catch (e) { /* do nothing */ }
+
+		// Check if we can safely use `fn.apply`.
+		if (hasApply) {
+			return function () {
+				fn.apply(console, arguments);
```
Comment (triblondon) [inline polyfills/console/polyfill.js:23]: Can't we use `Function.prototype.apply` here rather than checking for `apply` on `fn`?

**Comment by triblondon** [comment]
This is generally excellent. I'll come back to this in more detail when it's not Christmas Day...

File: polyfills/console/tests.js (hunk 1 above)
Comment (whitneyit) [inline polyfills/console/tests.js]: Oops

File: polyfills/console/polyfill.js (hunk 3 above)
Comment (whitneyit) [inline polyfills/console/polyfill.js:23]: I'm not sure. The angular guys saw fit to do it this way, I'm not sure of their reasoning nor their test cases.

File: polyfills/console/polyfill.js (hunk 2 above)
Comment (whitneyit) [inline polyfills/console/polyfill.js]: Updated

**Comment by whitneyit** [comment]
Thanks buddy, Merry Christmas to you too!

**Comment by triblondon** [comment]
Fails in all versions of IE:

```
Failed tests:
 - ie 11 (Sauce results: https://saucelabs.com/tests/2fd3457f1f454047b8ee9d7b9c7708c4)
    -> console
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=console
 - ie 10 (Sauce results: https://saucelabs.com/tests/026fdda06162445fbc8ce9aadd6da179)
    -> console
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=console
 - ie 9 (Sauce results: https://saucelabs.com/tests/30d270f42a6d4f50a31b04d30b42f5fe)
    -> console
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=console
 - ie 8 (Sauce results: https://saucelabs.com/tests/826b0a3b4e9e4ec5aafbc82356c9b612)
    -> console (timeout after 10000ms, 2 retries)
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=console (timeout after 10000ms, 2 retries)
 - ie 7 (Sauce results: https://saucelabs.com/tests/b56c27adb8274905af7c89af68078481)
    -> console (timeout after 10000ms, 2 retries)
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=console (timeout after 10000ms, 2 retries)
 - ie 6 (Sauce results: https://saucelabs.com/tests/47e86e02dd2b4aa3af8b87803c56fd1c)
    -> console (timeout after 10000ms, 2 retries)
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=console (timeout after 10000ms, 2 retries)
```

(except Edge)

**Comment by whitneyit** [comment]
Hey just getting back on this. I had a look through the SauceLabs screencast and it seems to die in the test when calling `throw()`.

**Comment by whitneyit** [comment]
Also, is it possible to add the `SAUCE_USERNAME` and `SAUCE_ACCESS_KEY` as environment variables in travis? The angular.js team put it in their [`.travis.yml`](https://github.com/angular/angular.js/blob/master/.travis.yml#L25-L26) file.

I'm not sure about the politics at FT, but this would make it easier ... for me at least :smile:

**Comment by whitneyit** [comment]
```
Failed tests:
 - ie 11 (Sauce results: https://saucelabs.com/tests/f193b141de4241ea87a45a69f6b4a9ef)
    -> console
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=console
 - ie 10 (Sauce results: https://saucelabs.com/tests/fa0c298c1ac84156b5bebecd1c297e8f)
    -> console
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=console
 - ie 9 (Sauce results: https://saucelabs.com/tests/9b7e652b00ed4f2fa9460a75db5c720d)
    -> console
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=console
 - ie 8 (Sauce results: https://saucelabs.com/tests/8e415c86c68648d984c1f3caaf5b6127)
    -> console
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=console
 - ie 7 (Sauce results: https://saucelabs.com/tests/8e13f2caf2fe45c1b3839ab61668e0e1)
    -> console
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=console
 - ie 6 (Sauce results: https://saucelabs.com/tests/99e4b3f867014ba0b5930a4ea157401a)
    -> console
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=console
```

**Comment by whitneyit** [comment]
So I fixed the issue with the reserved word `throw`. In other `tests,js` files, `not.to.throwException` is being used, updating the `.not.to.throw` with that solved the issue.

Also, each method and property is now split into its own folder.

**Comment by triblondon** [comment]
We're actually finding Sauce to be a bit of a problem, particularly in CI, and are looking at moving to Browserstack, but finding the time...

I'm a bit dubious of the capitalisation of `Throw`, is that supported by Expect? Anyway I'd suggest changing it to `throwException` - according to the Expect docs, it's a synonym of throw.

You then have an issue with your detect, because it is passing in IE11, which doesn't have all the console methods that you are testing for.  I made the detect more specific.

**Comment by triblondon** [comment]
See my PR from the local branch.

**Comment by whitneyit** [comment]
As you can see in my _updated_ earlier comment, I did change that. I was thinking that the `expect` being used was from Chai. I have updated the PR now

**Comment by whitneyit** [comment]
@triblondon do you think that we should polyfill `console.assert` to test the assertion and log it if `console.log` exists?

**Comment by whitneyit** [comment]
After a bit of playing around, the polyfill could be as simple as

``` js
console.assert = function (assertion, msg) {
  console.log(assertion, msg);
};
```

That's all Chrome seems to do. Thoughts?

**Comment by JakeChampion** [comment]
@whitneyit I like the recent comment you made -- https://github.com/whitneyit/polyfill-service/commit/52d377b36a5f7b4e7f6339419ef0fbd4186d318f

@triblondon What do you think about the separation of methods into their own folders?

**Comment by whitneyit** [comment]
I can rebase this on top of #569 if you like

**Comment by whitneyit** [comment]
@JakeChampion the separate folders also helps test individual methods that may not be supported. It solves the IE11 problem in #569 as well

**Comment by JakeChampion** [comment]
@whitneyit I've made a subsequent pr with your changes and removing the monolithic polyfill. I will run the tests locally as it seems SauceLabs is being flakey

**Comment by whitneyit** [comment]
I've tested this PR locally and it all passed

**Comment by whitneyit** [comment]
Wrote 35773 bytes to result file
Progress (browsers): 14 / 14 (0 browsers remaining, 0 test failures so far)
All jobs complete
Sauce tunnel stopped
travis_fold:end:Sauce test progress
Failed tests:
- None!
  Browsers tested with no failures: chrome/46, chrome/35, firefox/41, firefox/30, ie/12, ie/11, ie/10, ie/9, ie/8, ie/7, ie/6, safari/9, safari/8, android/4.4

**Comment by JakeChampion** [comment]
Did you run `grunt ci`?

**Comment by whitneyit** [comment]
I ran `grunt saucelabs:ci`

**Comment by JakeChampion** [comment]
Perfect work. I will assign the PR to Andrew and see if it can be merged in the morning (GMT)

**Comment by whitneyit** [comment]
Results are [here](https://gist.github.com/whitneyit/87931294df90e02c5db3)

**Comment by JakeChampion** [comment]
Thanks

**Comment by triblondon** [comment]
Happy to support the granular approach.

**Comment by triblondon** [comment]
Actually since my PR was merged and Jake's is an improvement on the architecture, the issue of adding the polyfill is done, so closing this.

**Pull request closed without merging**